### ISSUE TITLE
Decide what to do with Decision Services and multiple DRDs on the new DMN Editor

### ISSUE THREAD

**Issue by tiagobento** [apache/incubator-kie-issues#870]
As we evolve our new DMN Editor (#439), we want to come up with a conscious and definitive solution for supporting multiple DRDs while comprehensively managing the underlying DRG. Decision Services are complex enough even without multiple DRDs, but as DMN models can depict different portions of the DRG with many DRDs, we need to decide how we're going to deal with Decision Services and multiple DRDs.

The new DMN Editor already deals really well with single DRDs, making it easily to add/remove Decisions to/from Decision Services by dragging nodes in to / out of a Decision Service, and making it clear that these Decisions are "contained" by a Decision Service by giving them a thicker border.

<img width="1512" alt="image" src="https://github.com/apache/incubator-kie-issues/assets/1584568/591ed608-9c1a-45e4-a051-063f6ea0f493">

---

The DMN 1.5 specification, on section `6.2.5 Decision service`, only states that a Decision Service cannot be rendered in both collapsed and expanded form in the same DRD. Leaving a lot for interpretation and for implementers to decide.

It's also clear from the spec that Decision Services can be rendered in both expanded and collapsed form in multiple DRDs, but it is unclear what happens when one decision ceases to be part of a Decision Service, they'll continue to be visually enclosed by the Decision Service in other DRDs, although with normal border width in our new DMN Editor, it may be misleading. Leaving this burden for users to deal with when editing their DMNs is tedious and increases the cognitive load of dealing with Decision Services.

Now, using the statements we got from the DMN specification, and the DMN depicted above, let's discuss some of the difficulties of having Decision Services and multiple DRDs.

If I create a new `DRD B`, there are several things we can do in regards to `DS A`:

1. Allow adding `DS A` to it, without any of its contained Decisions.
> Easy to implement, but misleading as the Decision Service appears empty, when we know from the DRG that it is not. 

2. Allow adding `DS A` to it, bringing all its contained Decisions with it, automatically layout'ed inside `DS A`.
> Seems logical and very easy to use and understand, but will impose a lot of problems when modifying the DMN later. We'll discuss it below.

3. Allow adding `DS A` to it **only in collapsed form**, so none of its contained Decisions and dependent nodes will **ever** be displayed in `DRD B`
> Doesn't exactly conform to the specification recommendations, but does not violate any of the requirements. Not so easy to implement too, but allows for a smooth experience when editing the DMN later. We can even add an easy hook into the collapsed Decision Service that will take users to the DRD where the Decision Service is on its expanded form, allowing for easy editing. 
> This option is not free of downsides, though, as some of `DS A` dependencies might also be dependencies to nodes that depend on the `DS A`. More on this below.

4. Do not allow adding `DS A` to `DRD B` at all, as it is already depicted in `DRD A`.
> This will make it impossible to depict `DS A` in any way, making it impossible to show the full tree of a node that might depend on it via a Knowledge Requirement. Ultimately, this might make users ignore the multi-DRD capabilities of the new DMN Editor, as for some kinds of Decision models, it might be useless to separate only the portions of the Decision model that are not dependent on a Decision Service.

As you can see, each of those options lead to fundamentally different consequences which we need to consider in order to try and achieve the most efficient and pleasant experience for our users.

In my opinion, options 1 and 4 are out of the table, given the drawbacks they have. For options 2 and 3, I think we need to discuss more.

---

## Option 2 (Allow bringing a Decision Service into a DRD with all its contained Decisions automatically layout'ed inside it)

Consequences:
- A Decision Service can never be shown in its expanded form with a missing contained Decisions.
- We can't hide a Decision from a DRD if their contained Decision Service is depicted also depicted in the same DRD.
- When adding a new Decision to the DRG then adding it to an existing Decision Service, all DRDs that depict that Decision Service will need to be altered to automatically position the new Decision inside the Decision Service, to avoid misleading DRDs and weird behaviors when moving the Decision Service node around. This will potentially mess around too much with DRDs that are hidden in the background.
- The above is also true for moving Decisions out of the Decision Service. Hidden DRDs might continue showing the Decision inside the boundaries of the Decision Service, which might be misleading.

## Option 3 (Allow depicting a Decision Service in its expanded form in only one DRD, depicting it in its collapsed form in other DRDs)

Consequences:
- Limiting in terms of what users can do, and does not follow the suggestions of the DMN specification.
- Makes it easier for users to modify existing Decision Services, as their expanded form is limited to a single DRD, so users don't need to worry about messing with their other DRDs when modifying a Decision Service.
- If Decisions contained by a Decision Service are already depicted in a DRD, the Decision Service itself cannot be. And vice-versa.
- From our example above, this weird situation can happen in a new `DRD B`:
  <img width="400" alt="image" src="https://github.com/apache/incubator-kie-issues/assets/1584568/2e3007b7-6571-4719-9d09-833086dd801e">
  Even weirder, is this:
  <img width="400" alt="image" src="https://github.com/apache/incubator-kie-issues/assets/1584568/4f9c127f-d687-49c0-8bce-176841480b26">
  `D F` is a dependency of `D E`, which is encapsulated by `DS A`, that's in its collapsed form. No edges connect the two, though, making it look like `DS A` and `D F` do not have any kind of relationship. Let's call this `PROBLEM A`.
  Of course, adding `D D`, `D C`, or `D E` would not be allowed for this DRD.
- This situation below is also not so simple, as `D A` has a Knowledge Requirement to `DS A`. 
  <img width="400" alt="image" src="https://github.com/apache/incubator-kie-issues/assets/1584568/8d087477-2ef6-4825-b8b0-1e45533fc59e">
  Clicking the `...` on `D A` might even show `DS A` as a dependency, but we won't be able to show it on `DRD C`, since one or more of its contained Decisions are already present. Let's call this `PROBLEM B`.

---

As you can see, both options still have problems. IMHO, option 2 still puts too much in the users' shoulders and has the serious problem of tempering with DRDs that are not visible, but option 3 has `PROBLEM A` and `PROBLEM B`. Both are solvable, but we need to come up with good solutions to deal with them if we decide to go with option 3.

---

Note that I might have missed something during my analysis, so feedback is more than welcome! 

---

I'm really looking forward to hearing other people's opinions. Please do chime in!

**Comment by tiagobento** [comment]
- Btw, here's the DMN: 
  https://gist.github.com/tiagobento/02dd189779079728838185fbad890e6a

- And here's the link to import it directly into KIE Sandbox: 
  https://sandbox.kie.org/dev/#/import?url=https://gist.github.com/tiagobento/02dd189779079728838185fbad890e6a&branch=main

**Comment by jomarko** [comment]
Hi @tiagobento , let me share my thoughts about this topic. and my expectation, how the editor should work for this scenario.

----

I understand **DRDs** and **Decision Services** as two similar concepts, where both tries to simplify the model for users. However each does different simplification.

----

I understand **DRDs** more just like filtering of the nodes from the diagram. And each **DRD** as some stored filtered view of the main **DRG**. If I recall correctly, there can not be a node in **DRD** that wouldn't be in **DRG**. And I think **DRD** should contain:
- either whole **Decision Service** node
- or no **Decision Service** node

my reasoning for this is bellow.

----

I understand **Decision Services** as wrapper for some complex logic. These wrappers/decision services helps user to reuse some already created logic on more places, same as BKM nodes does. But from my point of view, it makes no sense to try reuse some wrapper/decision service on some other place and not include all its content. It would break not deterministic. I mean, if we once define Decision Service, it should always return the same results for the sme inputs. And this means Each occurrence of a decision Service node, is the same, containing all its **encapsulated** and **output** decisions.

So I think if a **Decision Service** node appears in any **DRD**, it should always represent the same logic. I do not have strong opinion if we use collapsed or expanded views to achieve this.

----

I think I speak about the same, what you describe as Option 3 but using other words?

**Comment by ljmotta** [comment]
I'm more inclined to option 2

I think that using a Decision Service should bring all nodes with it. As Jozef brought to the table, the idea of a Decision Service is to encapsulate a complex logic. Thinking about that, it's not reasonable to use a piece of this encapsulated logic (e.g. one decision node). With that in mind, I don't think it should be allowed to use nodes that were wrapped on it. 

Now, editing a Decision Service should reflect on all DRDs, even though it can be challenging. If it helps, we could limit editing this Decision Service on the DRD where it was defined (a mix with option 3).

Even though we bring all nodes with the Decision Service, I guess we should keep the option to collapse it. This would lead to option 3 `PROBLEM A`, but it could be solved by expanding the Decision Service when we have a node that has a relationship with it.

**Comment by tiagobento** [comment]
@ljmotta 

> _Thinking about that, it's not reasonable to use a piece of this encapsulated logic (e.g. one decision node). With that in mind, I don't think it should be allowed to use nodes that were wrapped on it._

I'm not sure I agree with this. The Decisions used as Output and Encapsulated Decisions in a Decision Service can 100% be reused by other parts of the DMN model without having to know that the Decision Service even exists. More than that, (and I totally forgot to bring this to the main discussion) **a single Decision can be a part of many different Decision Services.** This makes everything even more difficult, as if that happens for a DMN model, we won't ever be able to render the DRG with all Decision Services in expanded form.

---

My understanding is that Decision Services were introduced as a solution to reusability, not encapsulation.

---

If we allow collapsing Decision Service nodes (and I think we should...), we will have to implement Option 3. That means solving `PROBLEM A` and `PROBLEM B`.

**Comment by tiagobento** [comment]
@jomarko I agree with you that the representation of a Decision Service should be consistent in all DRDs.



> _I do not have strong opinion if we use collapsed or expanded views to achieve this._

Using only collapsed form is option 3, allowing expanded form is option 2.

**Comment by tiagobento** [comment]
@ljmotta and I had a very interesting conversation about this today... we went through some scenarios and simulated situations that can happen depending on what users want to do. We even discussed the situation where a Decision is contained by more than one Decision Service.

We kind of realized that options 2 and 3 complement each other, and we think we found a way to make all this complexity manageable. 

I'll try and summarize our findings here.

Premises:
- Any DRD node not contained by a Decision Service can be hidden or added from/to any DRD, including Decision Services themselves.
- Any Decision, including the ones contained by Decision Services, can be hidden or added from/to any DRD, except when one of its containing Decision Services is present in the same DRD, then this Decision is either inside one of its containing Decision Services (when in expanded form), or not showing at all (when collapsed). I.e., a Decision cannot appear outside of one of its containing Decision Services, even if all its containing Decision Services are collapsed.
- A Decision Service can either be I) Collapsed; or II) Expanded and showing all of its contained Decisions inside of its bounds. I.e., when a Decision Service is rendered in expanded form, all of its contained Decisions are added to the DRD, replicating other DRDs, or with auto-layout if no other DRD has an expanded form of this Decision Service.
- A Decision Service in expanded form must have exactly the same internal depiction in all DRDs.


### Short term actions
- If a Decision contained by a Decision Service is present on a DRD, no other Decision Service that contain that Decision can be expanded. I.e., on any DRD, only one of the Decision Services that contain a particular Decision can be in expanded form at a time.
- When a Decision Service is added to a DRD, all of its contained Decisions will be moved or added to the DRD, so that the internal depiction of that Decision Service is exactly the same as all other DRDs.
- Moving or changing the DRG of the Decisions contained by a Decision Service will replicate the changes to all other DRDs, so that all DRDs always have the same internal depiction of the Decision Service.
- Collapsing a Decision Service in one DRD does not alter other DRDs.
- When a Decision is removed from a Decision Service, this Decision is removed from all DRDs that contain that Decision Service.

### Medium/long term actions
- The normalization process introduced by #785 will be improved to make sure all DRDs have exactly the same depiction of a Decision Service and that all of the above applies. If there are no DRDs where all contained Decisions are depicted, we'll automatically lay them out inside the Decision Services that are in expanded form.
- When a Decision Service is collapsed, a new and very distinguishable edge will be rendered between nodes that have an direct or indirect relationship with one of the Decisions contained by the collapsed Decision Service. This new edge will also be used when there's a indirect relationship between nodes of the same DRD.

**Comment by jomarko** [comment]
Thank you for this comprehensive analysis of the Decision Services, I have couple of questions.
## Questions
- [ ] We mentioned already some scenarios, when adding a node into DRD will be not possible. I am interested how quickly we want to 'inform' user about this. Will be those 'not allowed' nodes disabled in the list of available nodes? Or all nodes will be always enabled but once dragged to the canvas, some validation will be fired and a flash message displayed if some rule will be violated?
- [ ] In the very last point some new Edge. Am I correct this is not part of the standard? Interested if some UX team can be contacted for some design recommendation can be contacted? Or does other decision modeling tools solve this already somehow?
- [ ] Unrelated to this ticket, but once some new DRD on https://sandbox.kie.org/dev, is there a way to delete DRD? I was not able to delete any DRD.

**Comment by tiagobento** [comment]
@jomarko 

- When adding a node to a DRD is not allowed, we will dim its opacity in the DRG / External nodes panels and won't make it draggable. No need for an error feedback, as we'll prevent the error from happening beforehand. 

- Yes, this would not be part of the standard and... we don't have a UX team :) Not sure how other Decision modelers go about this.

- Yes, you're right, we don't have the ability to delete a DRD yet, but Luiz and I are cooking something already for after #669, as this ticket will introduce the need to configure a DRD-specific flag.

---


Btw, thank you so much for being in the loop, I know this is complex and everything, but since you'll be working on #874, I think all this logic related to Decision Services and DRD will have to thoroughly tested.

**Comment by tiagobento** [comment]
Well, since I think we have a clear plan for this, I'm going to close this issue and add tasks to
- #625

**Comment by tiagobento** [comment]
For reference:

- #888
- #894
- #892
- #889
- #898
- #890
- #891
- #893
- #895
- #972
- https://github.com/apache/incubator-kie-issues/issues/1298
